chain/src/Pos/Chain/Block/Logic/Integrity.hs

```diff
@@ -189,16 +199,75 @@ verifyHeader pm VerifyHeaderParams {..} h =
+                    let slotLeader = addressHash $ mainHeader ^. mainHeaderLeaderKey
+                    in [  ( (slotLeader `elem` leaders)
+                          , sformat ("slot's leader, "%build%", is not an acceptable leader. acceptableLeaders: "%shown)
+                                slotLeader
+                                leaders)
+                        , ( (obftLeaderCanMint slotLeader)
+                          , sformat ("slot's leader, "%build%", has minted too many blocks in the past "%build%" slots.")
+                                slotLeader
+                                k)
+                        ]
+
+                OBFT ObftStrict ->
+                    let slotIndex = getSlotIndex $ siSlot $ mainHeader ^. headerSlotL
+                        slotLeader = leaders ^? ix (fromIntegral slotIndex)
+                        expectedSlotLeader = addressHash $ mainHeader ^. mainHeaderLeaderKey
+                    in [  ( (Just expectedSlotLeader == slotLeader)
+                          , sformat ("slot's leader, "%build%", is different from expected one, "%build%". slotIndex: "%build%", leaders: "%shown)
+                                slotLeader
+                                expectedSlotLeader
+                                slotIndex
+                                leaders)
+                        , ( (obftLeaderCanMint expectedSlotLeader)
+                          , sformat ("slot's leader, "%build%", has minted too many blocks in the past "%build%" slots.")
+                                slotLeader
+                                k)
+                        ]
+
+                Original ->
+                    let slotIndex = getSlotIndex $ siSlot $ mainHeader ^. headerSlotL
+                        slotLeader = leaders ^? ix (fromIntegral slotIndex)
+                        expectedSlotLeader = addressHash $ mainHeader ^. mainHeaderLeaderKey
+                    in [ ( (Just expectedSlotLeader == slotLeader)
+                        , sformat ("slot's leader, "%build%", is different from expected one, "%build%". slotIndex: "%build%", leaders: "%shown)
+                                slotLeader
+                                expectedSlotLeader
+                                slotIndex
+                                leaders)
+                        ]
+      where
+        -- Determine whether the leader is allowed to mint a block based on
+        -- whether blocksMintedByLeaderInLastKSlots <= floor (k * t)
+        obftLeaderCanMint leaderAddrHash =
+            (blocksMintedByLeaderInLastKSlots leaderAddrHash)
+                <= leaderMintThreshold
+        --
+        blocksMintedByLeaderInLastKSlots leaderAddrHash =
+            length $
+                filter (\lsi -> leaderAddrHash == (addressHash $ lsiLeaderPubkeyHash lsi))
+                       lastBlkSlots
+        --
+        leaderMintThreshold :: Int
+        leaderMintThreshold = floor $ (fromIntegral k :: Double) * t
```

Comment:
I guess it would just mean that the current leader would only be able to mint a block provided that they haven't minted _any_ blocks in the last `k` slots. 

This doesn't seem to be a problem during `OBFT` given enough potential leaders (due to the round-robin schedule) but tends to occasionally run into issues when verifying `Original` blocks created on FTS schedule with `OBFT` validation (this isn't typically done anyways).

Reply:
Ah right, so for OBFT with `k=2`, as long as we have 3 or more nodes, we should be fine 👍 

And yeah FTS could pose a problem, but if we're verifying mainnet we'd have a large enough `k` it would be fine.